### PR TITLE
Time splits

### PR DESCRIPTION
**Pull request by kspieks**
This PR enhances the list of features by adding time-based splits, which is somewhat popular within the field of cheminformatics. Hopefully adding this sampler encourages adoption of astartes!

This sampler require each datapoint to be associated with some date or datetime object. This PR implements the simplest approach in which the sampler simply sorts the datapoints by date or datetime and then assigns the oldest samples to the training set until that set is full, and then does the same for the validation and then test set. This follows the schematic from Figure 2 from [Feinberg et al.](https://pubs.acs.org/doi/full/10.1021/acs.jmedchem.9b02187) Since no clusters are created and it simply sorts indices, the `TimeBased` class actually behaves as an interpolation sampler behind the scenes. It's possible that future iterations could create clusters based on time and then have this act as an extrapolation sampler that assigns entire clusters to the respective sets; however, this seems ambigious as it's unclear what Δt would be appropriate to separate clusters based on time (e.g. if all datapoints are evenly space across time).

Just to document a few general discussion points:

1. Note that this sampler does not use the `random_state` argument. Since data is sorted chronologically, this is a deterministic operation, so there is no opportunity for randomness. One consequence of this is that there is no opportunity to create multiple folds and thus do cross-validation to obtain a mean and standard deviation across predictions from n models that were trained on n different splits of the data. However, when examining a few papers (now cited in the readme), it seems common to report results from just one split, and in fact the paper from [Robert P. Sheridan](https://pubs.acs.org/doi/full/10.1021/ci400084k) says, "Normally in the literature there would be multiple random-splits; however, because the other cross-validation schemes involve only a single split, we do that here", which confirms that the standard approach of using time-based splits is not concerned with creating multiple folds.
2. In general, time-based splits are not quite as rigorous as the other algorithmic samplers. The motivation for time-based splits is that the molecules under investigation typically change over time so eventually molecules will have structures or features that are likely not present in the molecules studied previously. However, this has a few ambiguities. First, at what time did the molecules start to become dissimilar? Are molecules studied now different from molecules studied x months ago? Second, if the answer is yes, how dissimilar are these new molecules from the ones looked at previously? There seems to be a need to quantify chemical similarity, which is handled by the other extrapolative samplers. Thus, it is likely more rigorous to directly split on chemical information rather than a proxy, such as a time, which may still leak substructures and/or reaction templates between the training, validation, and testing sets.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -60,13 +60,14 @@ There are two broad categories of sampling algorithms implemented in `astartes`:
...
 | Optimizable K-Dissimilarity Selection (OptiSim) | 'optisim' | Extrapolative | `n_clusters`, `max_subsample_size`, `distance_cutoff` | _custom implementation_ | Variation on [OptiSim](https://pubs.acs.org/doi/10.1021/ci025662h) for arbitrary-valued vectors. |
 | K-Means | 'kmeans' | Extrapolative | `n_clusters`, `n_init` | [`sklearn KMeans`](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html) | Passthrough to `sklearn`'s `KMeans`. |
 | Density-Based Spatial Clustering of Applications with Noise (DBSCAN) | 'dbscan' | Extrapolative | `eps`, `min_samples`, `algorithm`, `metric`, `leaf_size` | [`sklearn DBSCAN`](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html) | Passthrough to `sklearn`'s `DBSCAN`. |
 | Minimum Test Set Dissimilarity (MTSD) | ~ | ~ | _upcoming in_ `astartes` _v1.x_ | ~ | ~ |
 | Restricted Boltzmann Machine (RBM) | ~ | ~ | _upcoming in_ `astartes` _v1.x_ | ~ | ~ |
 | Kohonen Self-Organizing Map (SOM) | ~ | ~ | _upcoming in_ `astartes` _v1.x_ | ~ | ~ |
-| SPlit Method | ~ | ~ | _upcoming in_ `astartes` _v1.x_ | ~ | ~ |
```
Comment (JacksonBurns) [inline README.md:69]: This actually is the proper capitalization, I will update this.

File: astartes/samplers/extrapolation/scaffold.py [hunk 2]
```diff
@@ -28,6 +28,7 @@
 
 class Scaffold(AbstractSampler):
     def __init__(self, *args):
+        # ensure that X (i.e. args[0]) contains entries that are either a SMILES string or an RDKit Molecule
```
Comment (JacksonBurns) [inline astartes/samplers/extrapolation/scaffold.py:31]: Thanks for adding this as well!

File: README.md (hunk 1 above)
Comment (kspieks) [reply README.md:69]: Indeed this was my mistake. I should've checked the issues page on github. Thanks for correcting it!

**Comment by kspieks** [comment]
Thanks for adding the regression test for time based splits!

File: astartes/samplers/extrapolation/time_based.py
Comment (JacksonBurns) [inline astartes/samplers/extrapolation/time_based.py:1]: Great error checking, and every elegant implementation.

**Review by JacksonBurns — APPROVED** [review]
Thanks for adding this! Great discussion in the PR as well. I agree that time-based splitting seems to be a somewhat dubious proxy to more meaningful extrapolation approaches, but having it will hopefully bring more people into the fold.

LGTM

**Comment by kspieks** [comment]
Thanks for reviewing, improving, and merging!
